presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PlanNodeDecorrelator.java

```diff
@@ -59,98 +52,102 @@ public PlanNodeDecorrelator(PlanNodeIdAllocator idAllocator, Lookup lookup)
 
     public Optional<DecorrelatedNode> decorrelateFilters(PlanNode node, List<Symbol> correlation)
     {
-        PlanNodeSearcher filterNodeSearcher = searchFrom(node, lookup)
-                .where(FilterNode.class::isInstance)
-                .recurseOnlyWhen(isInstanceOfAny(ProjectNode.class, LimitNode.class));
-        List<FilterNode> filterNodes = filterNodeSearcher.findAll();
+        // TODO: when correlations list empty this should return immediately. However this isn't correct
+        // right now, because for nested subqueries correlation list is empty while there might exists usages
+        // of the outer most correlated symbols
```

Comment:
it is correct behaviour. It is because nested subquery is not correlated in such case, its input does not provide correlation symbols. 

After reading your explanation I still do not see a reason why it cannot return immediately in such case. `decorrelateFilters` will be called for outer correlated `LateralNode` anyway, right?